File: frontend/src/api/draft.ts
import { invoke } from "@tauri-apps/api/core";

import type { TemplateKind } from "../types/template";
import { DEFAULT_TEMPLATE_KIND } from "../types/template";

export async function getProjectDraft(projectId: string): Promise<string> {
  return invoke<string>("get_project_draft", { projectId });
}

export async function saveProjectDraft(
  projectId: string,
  draftText: string
): Promise<void> {
  await invoke<void>("save_project_draft", { projectId, draftText });
}

export async function generateStatusDraft(projectId: string): Promise<string> {
  return invoke<string>("generate_status_draft", { projectId });
}

export async function generateDraft(
  projectId: string,
  templateKind: TemplateKind
): Promise<string> {
  return invoke<string>("generate_draft", { projectId, templateKind });
}

export async function getProjectTemplateKind(
  projectId: string
): Promise<TemplateKind> {
  const kind = await invoke<string>("get_project_template_kind", { projectId });
  return (kind as TemplateKind) || DEFAULT_TEMPLATE_KIND;
}

export async function saveProjectTemplateKind(
  projectId: string,
  templateKind: TemplateKind
): Promise<void> {
  await invoke<void>("save_project_template_kind", {
    projectId,
    templateKind,
  });
}

File: src-tauri/src/commands/draft.rs
use tauri::State;

use crate::state::AppState;

#[tauri::command]
pub fn get_project_draft(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<String, String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service.get_draft(&project_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn save_project_draft(
    state: State<'_, AppState>,
    project_id: String,
    draft_text: String,
) -> Result<(), String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service
        .save_draft(&project_id, draft_text)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn generate_status_draft(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<String, String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service
        .generate_status_draft(&project_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn generate_draft(
    state: State<'_, AppState>,
    project_id: String,
    template_kind: String,
) -> Result<String, String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service
        .generate_draft(&project_id, &template_kind)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_project_template_kind(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<String, String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service
        .get_template_kind(&project_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn save_project_template_kind(
    state: State<'_, AppState>,
    project_id: String,
    template_kind: String,
) -> Result<(), String> {
    let service = state
        .project_service
        .lock()
        .map_err(|_| "Application state lock poisoned".to_string())?;

    service
        .save_template_kind(&project_id, &template_kind)
        .map_err(|e| e.to_string())
}
